# 1. Importing the libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
xls = pd.ExcelFile('Hash-Analytic-Python-Analytics-Problem-case-study-1.xlsx')
df1=pd.read_excel(xls,'Existing employees')
df2=pd.read_excel(xls,'Employees who have left')

In [3]:
df1.head()

Emp ID  satisfaction_level  last_evaluation  number_project  \
0    2001                0.58             0.74               4   
1    2002                0.82             0.67               2   
2    2003                0.45             0.69               5   
3    2004                0.78             0.82               5   
4    2005                0.49             0.60               3   

   average_montly_hours  time_spend_company  Work_accident  \
0                   215                   3              0   
1                   202                   3              0   
2                   193                   3              0   
3                   247                   3              0   
4                   214                   2              0   

   promotion_last_5years   dept salary  
0                      0  sales    low  
1                      0  sales    low  
2                      0  sales    low  
3                      0  sales    low  
4                      0  sales    low

In [4]:
df2.head()

Emp ID  satisfaction_level  last_evaluation  number_project  \
0       1                0.38             0.53               2   
1       2                0.80             0.86               5   
2       3                0.11             0.88               7   
3       4                0.72             0.87               5   
4       5                0.37             0.52               2   

   average_montly_hours  time_spend_company  Work_accident  \
0                   157                   3              0   
1                   262                   6              0   
2                   272                   4              0   
3                   223                   5              0   
4                   159                   3              0   

   promotion_last_5years   dept  salary  
0                      0  sales     low  
1                      0  sales  medium  
2                      0  sales  medium  
3                      0  sales     low  
4                      0  sales     low

# 2. Combining data of employees who left and existing employees in order to study the status visually

In [5]:
df1['status']=0
df2['status']=1
frames=[df1,df2]
df=pd.concat(frames, axis=0, ignore_index=True)
df=df.sort_values(by='Emp ID', ascending=True)

In [6]:
df

Emp ID  satisfaction_level  last_evaluation  number_project  \
11428       1                0.38             0.53               2   
11429       2                0.80             0.86               5   
11430       3                0.11             0.88               7   
11431       4                0.72             0.87               5   
11432       5                0.37             0.52               2   
...       ...                 ...              ...             ...   
14994   14995                0.40             0.57               2   
14995   14996                0.37             0.48               2   
14996   14997                0.37             0.53               2   
14997   14998                0.11             0.96               6   
14998   14999                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  \
11428                   157                   3              0   
11429                   262                   6              0   
11430                   272                   4              0   
11431                   223                   5              0   
11432                   159                   3              0   
...                     ...                 ...            ...   
14994                   151                   3              0   
14995                   160                   3              0   
14996                   143                   3              0   
14997                   280                   4              0   
14998                   158                   3              0   

       promotion_last_5years     dept  salary  status  
11428                      0    sales     low       1  
11429                      0    sales  medium       1  
11430                      0    sales  medium       1  
11431                      0    sales     low       1  
11432                      0    sales     low       1  
...                      ...      ...     ...     ...  
14994                      0  support     low       1  
14995                      0  support     low       1  
14996                      0  support     low       1  
14997                      0  support     low       1  
14998                      0  support     low       1  

[14999 rows x 11 columns]

# 3. Encoding categorical data

In [7]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [8]:
enc=LabelEncoder()

In [9]:
count=0
for i in df1.columns[0:]:
    if df1[i].dtype=='object':
        enc.fit(df1[i])
        df1[i]=enc.transform(df1[i])
        count+=1

In [10]:
count

2

In [11]:
enc1 = OneHotEncoder(handle_unknown='ignore')

In [12]:
enc_df1 = pd.DataFrame(enc1.fit_transform(df[['dept']]).toarray())
enc_df2 = pd.DataFrame(enc1.fit_transform(df[['salary']]).toarray())

In [13]:
enc_df1

0    1    2    3    4    5    6    7    8    9
0      0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
1      0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
2      0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
3      0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
4      0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...
14994  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
14995  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
14996  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
14997  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
14998  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0

[14999 rows x 10 columns]

In [14]:
enc_df1=enc_df1.drop([0], axis=1)

In [15]:
enc_df1.shape


(14999, 9)

In [16]:
enc_df2

0    1    2
0      0.0  1.0  0.0
1      0.0  0.0  1.0
2      0.0  0.0  1.0
3      0.0  1.0  0.0
4      0.0  1.0  0.0
...    ...  ...  ...
14994  0.0  1.0  0.0
14995  0.0  1.0  0.0
14996  0.0  1.0  0.0
14997  0.0  1.0  0.0
14998  0.0  1.0  0.0

[14999 rows x 3 columns]

In [17]:
enc_df2=enc_df2.drop([0], axis=1)

In [18]:
enc_df2.shape

(14999, 2)

In [19]:
df.shape

(14999, 11)

In [20]:
df=pd.concat([df,enc_df1,enc_df2], axis=1, join='outer')

In [21]:
df.shape

(14999, 22)

In [23]:
df.head()

Emp ID  satisfaction_level  last_evaluation  number_project  \
0    2001                0.58             0.74               4   
1    2002                0.82             0.67               2   
2    2003                0.45             0.69               5   
3    2004                0.78             0.82               5   
4    2005                0.49             0.60               3   

   average_montly_hours  time_spend_company  Work_accident  \
0                   215                   3              0   
1                   202                   3              0   
2                   193                   3              0   
3                   247                   3              0   
4                   214                   2              0   

   promotion_last_5years   dept salary  ...    2    3    4    5    6    7  \
0                      0  sales    low  ...  0.0  0.0  0.0  0.0  0.0  1.0   
1                      0  sales    low  ...  0.0  0.0  0.0  0.0  0.0  1.0   
2                      0  sales    low  ...  0.0  0.0  0.0  0.0  0.0  1.0   
3                      0  sales    low  ...  0.0  0.0  0.0  0.0  0.0  1.0   
4                      0  sales    low  ...  0.0  0.0  0.0  0.0  0.0  1.0   

     8    9    1    2  
0  0.0  0.0  1.0  0.0  
1  0.0  0.0  0.0  1.0  
2  0.0  0.0  0.0  1.0  
3  0.0  0.0  1.0  0.0  
4  0.0  0.0  1.0  0.0  

[5 rows x 22 columns]

In [24]:
df.shape

(14999, 22)

In [26]:
X=df.iloc[:, 0:]
y=df.iloc[:,10]
X.drop(['Emp ID', 'Work_accident', 'salary', 'dept', 'status'], axis=1, inplace=True)

In [27]:
X.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.58             0.74               4                   215   
1                0.82             0.67               2                   202   
2                0.45             0.69               5                   193   
3                0.78             0.82               5                   247   
4                0.49             0.60               3                   214   

   time_spend_company  promotion_last_5years    1    2    3    4    5    6  \
0                   3                      0  0.0  0.0  0.0  0.0  0.0  0.0   
1                   3                      0  0.0  0.0  0.0  0.0  0.0  0.0   
2                   3                      0  0.0  0.0  0.0  0.0  0.0  0.0   
3                   3                      0  0.0  0.0  0.0  0.0  0.0  0.0   
4                   2                      0  0.0  0.0  0.0  0.0  0.0  0.0   

     7    8    9    1    2  
0  1.0  0.0  0.0  1.0  0.0  
1  1.0  0.0  0.0  0.0  1.0  
2  1.0  0.0  0.0  0.0  1.0  
3  1.0  0.0  0.0  1.0  0.0  
4  1.0  0.0  0.0  1.0  0.0

In [28]:
X.shape

(14999, 17)

## 4. Estimating the accuracy

In [29]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

In [31]:
X_train.shape , X_test.shape , y_train.shape, y_test.shape

((11999, 17), (3000, 17), (11999,), (3000,))

In [32]:
clf_tree = DecisionTreeClassifier(criterion='entropy', max_depth=3, 
                                  random_state=17)

In [33]:
clf_tree.fit(X_train, y_train)


DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=3, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=17, splitter='best')

In [34]:
y_pred=clf_tree.predict(X_test)

In [35]:
from sklearn.metrics import accuracy_score

In [36]:
accuracy_score(y_test, y_pred)

0.9436666666666667

### Thus, the proposed model "DECISION TREE ALGORITHM" can predict the employee churn with an accuracy of 94.37%